File: app/javascript/channels/order_channel.js
import consumer from "channels/consumer"

export class OrderChannel {
   static subscription = null
   constructor(controller) {
      this.controller = controller;
      this.subscribe();
   }

   connected() {
      console.log("connected to OrderChannel");
   }

   disconnected() {
   }

   received(data) {
      // called when there's incoming data on the websocket for this subscription
      console.log("received data:", data);
      if (data.form_html) {
         this.controller.updateForm(data.form_html);
      }

      if (data.html) {
        this.controller.addItem(data);
      }
      else if (data.quantity == 0) {
         this.controller.deleteItem(data);
      }
      else {
         this.controller.updateItem(data);
      }

      this.controller.updateTotalPrice(data.total_price);
   }

   changeItemQuantity(product_id, quantity) {
      console.log("updating order item:", product_id, quantity);
      const orderItem = { product_id: product_id, quantity: quantity };
      OrderChannel.subscription.perform('change_item_quantity_by', orderItem);
   }

   setItemQuantity(product_id, quantity) {
      console.log("set item quantity", product_id, quantity);
      const orderItem = { product_id: product_id, quantity: quantity };
      OrderChannel.subscription.perform('set_item_quantity_to', orderItem);
   }

   subscribe() {
      if (OrderChannel.subscription) {
         console.log("subscription already exists");
         return;
      }
      OrderChannel.subscription = consumer.subscriptions.create("OrderChannel", {
         connected: this.connected.bind(this),
         disconnected: this.disconnected.bind(this),
         received: this.received.bind(this),
      })
   }

   unsubscribe() {
      if (OrderChannel.subscription) {
         OrderChannel.subscription.unsubscribe();
         OrderChannel.subscription = null;
         console.log("OrderChannel unsubscribed");
      }
   }
}

File: app/channels/order_channel.rb
class OrderChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    Rails.logger.info("Subscribed to OrderChannel for user #{current_user.id}")
    stream_from "cart_#{current_user.id}"
  end

  def change_item_quantity_by(data)
    modify_item_quantity(data) do |current_quantity, quantity|
      current_quantity + quantity
    end
  end

  def set_item_quantity_to(data)
    modify_item_quantity(data) do |current_quantity, quantity|
      quantity
    end
  end


  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  private
  def modify_item_quantity(data)
    # Find cart and update the order item quantity
    cart = Order.cart_for_user(current_user).first_or_initialize
    product_id = data["product_id"]
    quantity = data["quantity"].to_i
    item = cart.order_items.where(product_id: product_id).first_or_initialize

    quantity = yield(item.quantity, quantity)
    item.quantity = quantity
    item.price = item.product.price

    if item.persisted?
      if item.save and cart.save
        # Broadcast the updated cart to client
        Rails.logger.info("Updating existing order item for product_id: #{product_id}")
        ActionCable.server.broadcast(
          "cart_#{current_user.id}",
          {
            product_id: product_id,
            quantity: quantity,
            total_price: cart.total_price
          }
        )
        item.destroy if item.quantity <= 0
      end
    else
      # Create a new item
      if item.save and cart.save
        # Broadcast the updated cart to client
        Rails.logger.info("Creating new order item for product_id: #{product_id}")
        ActionCable.server.broadcast(
          "cart_#{current_user.id}",
          {
            html: ApplicationController.render(
              partial: "orders/item",
              locals: { item: item, cart: true }
            ),
            total_price: cart.total_price
          }
        )
      end
    end
  end
end
